### PR TITLE
[PLUGIN-1762][PLUGIN-1764]Access Token support Added and Macro added for refresh token fields.

### PR DESCRIPTION
**Pull request by vikasrathee-cs**
[PLUGIN-1762] Macros added for oauth fields- clientId, clientSecret and refresh token.
[PLUGIN-1764] Access token field added that can be used with oauthAccessToken macro.

### CODE REVIEW

File: docs/GoogleDrive-batchsink.md [hunk 1]
```diff
@@ -43,6 +43,11 @@ Make sure that:
 OAuth2 client credentials can be generated on Google Cloud
 [Credentials Page](https://console.cloud.google.com/apis/credentials)
 
+**OAuth Method:** The method used to get OAuth access tokens. The oauth access token can be directly provided,
```
Comment (itsankit-google) [inline docs/GoogleDrive-batchsink.md]: what is `OAuthMethod`, can you give some examples about method?

Is it some standard practice?

File: widgets/GoogleDrive-batchsource.json [hunk 2]
```diff
@@ -488,6 +515,30 @@
...
+        {
+          "name": "oauthMethod",
+          "type": "property"
+        }
+      ]
+    },
+    {
+      "name": "Authenticate with OAuth2 Access Token",
+      "condition": {
+        "expression": "oauthMethod == 'ACCESS_TOKEN' && authType == 'oAuth2'"
```
Comment (itsankit-google) [inline widgets/GoogleDrive-batchsource.json]: when we upgrade older pipeline using `oAuth2`, it will not have `oAuthMethod` set, did we test this scenario?

I think the expression should be `oAuthMethod != 'REFRESH_TOKEN' && authType == 'oAuth2'`, please test it for older pipelines

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 3]
```diff
@@ -76,21 +79,38 @@ public abstract class GoogleAuthBaseConfig extends PluginConfig {
   @Macro
   protected String serviceAccountType;
 
+  @Macro
+  @Nullable
+  @Name(OAUTH_METHOD)
+  @Description("The method used to get OAuth access tokens. "
+    + "The oauth access token can be directly provided, "
+    + "or a client id, client secret, and refresh token can be provided.")
+  private String oauthMethod;
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: nit: `oAuthMethod`

File: docs/GoogleDrive-batchsink.md (hunk 1 above)
Comment (vikasrathee-cs) [reply docs/GoogleDrive-batchsink.md]: This can be used where user wants to use oauthAccessToken macro. Added the same in doc.

File: widgets/GoogleDrive-batchsource.json (hunk 2 above)
Comment (vikasrathee-cs) [reply widgets/GoogleDrive-batchsource.json]: we have do a null check here in case of upgrade, added the same

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java (hunk 3 above)
Comment (vikasrathee-cs) [reply src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: done

File: docs/GoogleDrive-batchsink.md (hunk 1 above)
Comment (itsankit-google) [reply docs/GoogleDrive-batchsink.md]: The inclusion of the additional line appears to be redundant.

Just keep it like this in the doc:
```
The method used to get OAuth access tokens. Users have the option to either directly provide the OAuth access token or supply a client ID, client secret, and refresh token.
```

File: docs/GoogleDrive-batchsink.md [hunk 4]
```diff
@@ -43,6 +43,12 @@ Make sure that:
...
+or a client id, client secret, and refresh token can be provided. Access Token can be used if user wants to
+provide access token directly or oauthAccessToken macro.
+
+**Access Token:** Short lived access token for connect.
```
Comment (itsankit-google) [inline docs/GoogleDrive-batchsink.md]:
```
Short lived access token used for connecting.
```

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 5]
```diff
@@ -76,21 +79,38 @@ public abstract class GoogleAuthBaseConfig extends PluginConfig {
   @Macro
   protected String serviceAccountType;
 
+  @Macro
+  @Nullable
+  @Name(OAUTH_METHOD)
+  @Description("The method used to get OAuth access tokens. "
+    + "The oauth access token can be directly provided, "
+    + "or a client id, client secret, and refresh token can be provided.")
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: Similar comment here:
```
The method used to get OAuth access tokens. Users have the option to either directly provide the OAuth access token or supply a client ID, client secret, and refresh token.
```

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 6]
```diff
@@ -76,21 +79,38 @@ public abstract class GoogleAuthBaseConfig extends PluginConfig {
...
   @Nullable
+  @Macro
   @Name(REFRESH_TOKEN)
   @Description("OAuth2 refresh token.")
   private String refreshToken;
 
+  @Nullable
+  @Macro
+  @Name(ACCESS_TOKEN)
+  @Description("Short lived access token for connect.")
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]:
```
Short lived access token used for connecting.
```

File: widgets/GoogleDrive-batchsource.json (hunk 2 above)
Comment (itsankit-google) [reply widgets/GoogleDrive-batchsource.json]: yeah I saw a null check in backend for upgrade this comment was more about UI

File: widgets/GoogleSheets-batchsink.json [hunk 7]
```diff
@@ -234,7 +261,33 @@
     {
       "name": "Authenticate with OAuth2",
       "condition": {
-        "expression": "authType == 'oAuth2'"
+        "property": "authType",
+        "operator": "equal to",
+        "value": "oAuth2"
```
Comment (itsankit-google) [inline widgets/GoogleSheets-batchsink.json:266]: why this change?

Comment (vikasrathee-cs) [reply widgets/GoogleSheets-batchsink.json:266]: this filter is there in all 4 plugins, made them similar

File: widgets/GoogleSheets-batchsink.json [hunk 8]
```diff
@@ -234,7 +261,33 @@
...
+        {
+          "name": "accessToken",
+          "type": "property"
+        }
+      ]
+    },
+    {
+      "name": "Authenticate with OAuth2 Refresh Token",
+      "condition": {
+        "expression": "oAuthMethod == null || (oAuthMethod != 'ACCESS_TOKEN' && authType == 'oAuth2')"
```
Comment (itsankit-google) [inline widgets/GoogleSheets-batchsink.json:290]: this expression can be simplified:

```
oAuthMethod != 'REFRESH_TOKEN' && authType == 'oAuth2'
```

File: widgets/GoogleDrive-batchsource.json (hunk 2 above)
Comment (vikasrathee-cs) [reply widgets/GoogleDrive-batchsource.json]: on the UI only added a null check now

File: widgets/GoogleSheets-batchsink.json (hunk 8 above)
Comment (vikasrathee-cs) [reply widgets/GoogleSheets-batchsink.json:290]: this is not working as oAuthMethod tag is not coming in json at the time of upgrade, after that only I added null check

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java [hunk 9]
```diff
@@ -54,6 +54,9 @@ public abstract class GoogleAuthBaseConfig extends PluginConfig {
   public static final String SERVICE_ACCOUNT_FILE_PATH = "filePath";
   public static final String SERVICE_ACCOUNT_JSON = "JSON";
   public static final String SCHEMA = "schema";
+  public static final String ACCESS_TOKEN = "accessToken";
+  public static final String ACCESS_TOKEN_LABEL = "Access token";
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: IIUC label is picked from widgets json, there it is specified as `Access Token` right?

Comment (itsankit-google) [reply src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: similar problem with `REFRESH_TOKEN_LABEL` on line 49.

File: widgets/GoogleSheets-batchsink.json (hunk 8 above)
Comment (itsankit-google) [reply widgets/GoogleSheets-batchsink.json:290]: I see, thanks for the info.

File: src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java (hunk 9 above)
Comment (vikasrathee-cs) [reply src/main/java/io/cdap/plugin/google/common/GoogleAuthBaseConfig.java]: This is used to just to form error message, I changed all the labels to have uniformity.

File: src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java [hunk 10]
```diff
@@ -282,7 +282,9 @@ private boolean shouldGetSchema() {
     return !containsMacro(SHEETS_TO_PULL) && !containsMacro(SHEETS_IDENTIFIERS) &&
       !containsMacro(COLUMN_NAMES_SELECTION) && !containsMacro(CUSTOM_COLUMN_NAMES_ROW) &&
       !containsMacro(LAST_DATA_COLUMN) && !containsMacro(NAME_SERVICE_ACCOUNT_TYPE) &&
-      !containsMacro(ACCOUNT_FILE_PATH) && !containsMacro(NAME_SERVICE_ACCOUNT_JSON);
+      !containsMacro(ACCOUNT_FILE_PATH) && !containsMacro(NAME_SERVICE_ACCOUNT_JSON) &&
+      !containsMacro(CLIENT_ID) && !containsMacro(CLIENT_SECRET) &&
+      !containsMacro(REFRESH_TOKEN) && !containsMacro(ACCESS_TOKEN);
```
Comment (itsankit-google) [inline src/main/java/io/cdap/plugin/google/sheets/source/GoogleSheetsSourceConfig.java]: we should also include `oAuthMethod` here since it can also be macro.

**Review by itsankit-google — COMMENTED** [review]
LGTM, Please squash commits so that we can merge the PR.
